### DISCUSSION TITLE
Renovate bot creating too many Merge Requests despite grouping configuration

### DISCUSSION THREAD

**Discussion by badrshs** [renovatebot/renovate#31424] [Request Help]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitLab

### Please tell us more about your question or problem

I'm trying to configure Renovate bot for my project, but I'm having trouble getting it to create the exact number of Merge Requests (MRs) I want. My project uses both npm for frontend dependencies and NuGet for backend dependencies.

## Current Situation

With my current configuration, Renovate is creating 68 MRs, which is far more than I need. Here's a snippet of what I'm seeing:

```
With your current configuration, Renovate will create 80 Merge Requests:

► Pin dependencies
► Update angular monorepo angular monorepo (patch)
► Update angular-cli monorepo angular-cli monorepo (patch)
► Update aws-sdk-net monorepo aws-sdk-net monorepo (patch)
► Update backend-dependencies backend-dependencies (patch)
► Update dependency datatables.net-bs5
► Update dependency jasmine-core
► Update dependency typescript-eslint
...
[many more similar entries]
...
► Update angular monorepo angular monorepo (major)
```

## Desired Outcome

I want to significantly reduce the number of MRs to just 5:

1. Frontend patch updates
2. Frontend minor updates
3. Frontend major updates
4. Backend patch and minor updates combined
5. Backend major updates

## My Current Configuration

Here's my current Renovate configuration:

```json
{
  "extends": ["config:recommended"],
  "packageRules": [
    {
      "matchManagers": ["npm"],
      "matchUpdateTypes": ["patch"],
      "groupName": "frontend-patch-dependencies",
      "groupSlug": "frontend-patch-updates"
    },
    {
      "matchManagers": ["npm"],
      "matchUpdateTypes": ["minor"],
      "groupName": "frontend-minor-dependencies",
      "groupSlug": "frontend-minor-updates"
    },
    {
      "matchManagers": ["npm"],
      "matchUpdateTypes": ["major"],
      "groupName": "frontend-major-dependencies",
      "groupSlug": "frontend-major-updates"
    },
    {
      "matchManagers": ["nuget"],
      "matchUpdateTypes": ["patch", "minor"],
      "groupName": "backend-patch-minor-dependencies",
      "groupSlug": "backend-patch-minor-updates"
    },
    {
      "matchManagers": ["nuget"],
      "matchUpdateTypes": ["major"],
      "groupName": "backend-major-dependencies",
      "groupSlug": "backend-major-updates"
    }
  ],
  "timezone": "UTC",
  "prHourlyLimit": 2,
  "prConcurrentLimit": 5,
  "separateMultipleMajor": false,
  "separateMinorPatch": false,
  "separateMajorMinor": false,
  "commitMessageExtra": "{{groupName}}",
  "commitBodyTable": true,
  "lockFileMaintenance": {
    "enabled": true,
    "automerge": false
  },
  "rangeStrategy": "pin",
  "prCreation": "immediate",
  "labels": ["dependencies"],
  "dependencyDashboard": true
}
```

I also tried 

    {
      "extends": ["config:recommended"],
      "packageRules": [
      {
        "groupName": "all dependencies",
        "groupSlug": "all",
        "matchPackagePatterns": [
          "*"
        ],
        "matchUpdateTypes": [
          "major",
          "minor",
          "patch"
        ]
      }
    ],
      "timezone": "UTC",
      "prHourlyLimit": 2,
      "prConcurrentLimit": 5,
      "separateMultipleMajor": false,
      "separateMinorPatch": false,
      "separateMajorMinor": false,
      "commitMessageExtra": "{{groupName}}",
      "commitBodyTable": true,
      "lockFileMaintenance": {
        "enabled": true,
        "automerge": false
      },
      "rangeStrategy": "pin",
      "prCreation": "immediate",
      "labels": ["dependencies"],
      "dependencyDashboard": true
    }

I assumed the one above should create only 1 MR but it did 

[![enter image description here][1]][1]
## Question

How can I modify my Renovate configuration to create exactly 5 MRs as described in the "Desired Outcome" section? I've tried various grouping strategies, but I'm still getting too many individual MRs.

Any help in refining this configuration would be greatly appreciated!

  [1]: https://i.sstatic.net/M4UakYpB.png

**Comment by rarkins** [comment]
The monorepo preset rules coming from `config:recommended` are taking precedence over your own rules. Where/how are you defining your config? If it's in `renovate.json` then it's unexpected that preset rules would take precedence. If you're defining them in `config.js` then it's expected, because `extends` takes priority over global config.

**Reply by badrshs** [reply]
I'm was doing it from renovate.json  , and I'm not using config.js at all  

After long search it seems working now with the following but I don't feel it's 100% correct as I'm not using  `config:recommended`

```
{
  "extends": ["mergeConfidence:all-badges"],
  "packageRules": [
    {
      "matchManagers": ["npm"],
      "matchPackagePatterns": [".*"],
      "matchUpdateTypes": [ "patch","pin", "digest", "pinDigest"],
      "groupName": "front-end-patch-dependencies",
      "groupSlug": "front-end-patch-updates"
    },
    {
      "matchManagers": ["npm"],
      "matchPackagePatterns": [".*"],
      "matchUpdateTypes": ["minor"],
      "groupName": "front-end-minor-dependencies",
      "groupSlug": "front-end-minor-updates"
    },
    {
      "matchManagers": ["npm"],
      "matchPackagePatterns": [".*"],
      "matchUpdateTypes": ["major"],
      "groupName": "front-end-major-dependencies",
      "groupSlug": "front-end-major-updates"
    },
    {
      "matchManagers": ["nuget"],
      "matchUpdateTypes": [ "patch","pin", "minor", "digest", "lockFileMaintenance", "pinDigest"],
      "groupName": "backend-minor-patch-dependencies",
      "groupSlug": "backend-minor-patch-updates"
    },
    {
      "matchManagers": ["nuget"],
      "matchUpdateTypes": ["major"],
      "enabled": true,
      "groupName": "backend-major-dependencies",
      "groupSlug": "backend-major-updates-"
    }
  ],
  "hostRules": [
    {
      "matchHost": "https://nuget.x.local/nuget/x.NuGet.Server/v3/index.json",
      "hostType": "nuget",
      "token": "{{NUGET_API_KEY}}"
    }
  ],
  "nuget": {
    "registryUrls": [
      "https://nuget.x.local/nuget/x.NuGet.Server/v3/index.json",
      "https://api.nuget.org/v3/index.json"
    ]
  },
  "timezone": "UTC",
  "ignoreSchedules": ["*"],
  "schedule": ["at any time"],
  "prHourlyLimit": 0,
  "prConcurrentLimit": 0,
  "separateMultipleMajor": false,
  "separateMinorPatch": false,
  "separateMajorMinor": false,
  "commitMessageExtra": "{{groupName}}",
  "commitBodyTable": true,
  "rangeStrategy": "pin",
  "prCreation": "immediate",
  "labels": ["dependencies"],
  "dependencyDashboard": true 
}
```

any suggestion or something doesn't make sense to you here ? mainly the custom nuget customization   😓

**Reply by rarkins** [reply]
Better to use packageRules with matchDatasources than the nuget {} approach
